command/registry.go

```diff
@@ -0,0 +1,173 @@
+// for the command
+func TrySetupHCP(cfg packer.Handler) hcl.Diagnostics {
+	switch cfg := cfg.(type) {
+	case *hcl2template.PackerConfig:
+		return setupRegistryForPackerConfig(cfg)
+	case *CoreWrapper:
+		return setupRegistryForPackerCore(cfg)
+	}
+
+	return hcl.Diagnostics{
+		&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "unknown Handler type",
+			Detail: "SetupRegistry called with an unknown Handler. " +
+				"This is a Packer bug and should be brought to the attention " +
+				"of the Packer team, please consider opening an issue for this.",
+		},
+	}
+}
+
+func setupRegistryForPackerConfig(pc *hcl2template.PackerConfig) hcl.Diagnostics {
+	var diags hcl.Diagnostics
+
+	hasHCP := false
+
+	for _, build := range pc.Builds {
+		if build.HCPPackerRegistry != nil {
+			hasHCP = true
+		}
+	}
+
+	if hasHCP && len(pc.Builds) > 1 {
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Multiple " + buildLabel + " blocks",
+			Detail: fmt.Sprintf("For Packer Registry enabled builds, only one " + buildLabel +
+				" block can be defined. Please remove any additional " + buildLabel +
+				" block(s). If this " + buildLabel + " is not meant for the Packer registry please " +
+				"clear any HCP_PACKER_* environment variables."),
+		})
+		return diags
+	}
+
+	if env.IsPAREnabled() {
+		hasHCP = true
+	}
+
+	if !hasHCP {
+		return diags
+	}
+
+	var err error
+	pc.Bucket, err = registry.NewBucketWithIteration(registry.IterationOptions{
+		TemplateBaseDir: pc.Basedir,
+	})
+
+	if err != nil {
+		diags = append(diags, &hcl.Diagnostic{
+			Summary:  "Unable to create a valid bucket object for HCP Packer Registry",
+			Detail:   fmt.Sprintf("%s", err),
+			Severity: hcl.DiagError,
+		})
+
+		return diags
+	}
+
+	pc.Bucket.LoadDefaultSettingsFromEnv()
+
+	for _, build := range pc.Builds {
+		build.HCPPackerRegistry.WriteToBucketConfig(pc.Bucket)
+
+		// If at this point the bucket.Slug is still empty,
+		// last try is to use the build.Name if present
+		if pc.Bucket.Slug == "" && build.Name != "" {
+			pc.Bucket.Slug = build.Name
+		}
+
+		// If the description is empty, use the one from the build block
+		if pc.Bucket.Description == "" {
+			pc.Bucket.Description = build.Description
+		}
+
+		for _, source := range build.Sources {
+			pc.Bucket.RegisterBuildForComponent(source.String())
+		}
+	}
+
+	hcpData := map[string]cty.Value{}
+
+	vals, diags := pc.Datasources.Values()
+	if diags != nil {
+		return diags
+	}
+
+	// If we don't have any HCP image datasource, no need to try and get them
+	dsVals, ok := vals[hcpImageDatasouceType]
+	if !ok {
+		return nil
+	}
+
+	err = gocty.FromCtyValue(dsVals, &hcpData)
+	if err != nil {
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid HCP datasources",
+			Detail:   fmt.Sprintf("Failed to decode hcp_packer_image datasources: %s", err),
+		})
+		return diags
```

Comment:
Making this a hard fail seems odd to me since this is work being done behind the scenes to identify potential source images. If we can't get this information for whatever reason we should not prevent a build. At least that's how I see it. Especially since there is a background job on the HCP Packer side that can help make this association @sylviamoss thoughts? 

That said, @lbajolet-hashicorp how would we end up in this state where dsVals is not a valid data source? Is there something we can do further up to ensure we don't end up with an invalid data source?

Reply:
Quite frankly if we reach this code, to me it's a programming error on our side, that could even become a `panic`.